### PR TITLE
Phase 6: Langevin Gradient Noise (SGLD) — Stochastic Exploration for Lion

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Lion is a sign-based optimizer: every gradient update is quantized to ±1 per parameter. This eliminates the inherent stochasticity that SGD/Adam get for free from minibatch gradient variance. Once Lion's sign has converged for a weight, it stays locked unless the gradient distribution changes meaningfully.

Adding calibrated Gaussian noise to parameters **after** each Lion step implements Stochastic Gradient Langevin Dynamics (SGLD) — a Bayesian optimization technique that encourages exploration of flat minima in the loss landscape. We anneal the noise to zero before EMA starts (~epoch 140), so EMA averages over a stabilized, noise-free basin rather than noisy checkpoints.

**Why this might help p_in:** L1 surface MAE on in-distribution single-foil samples (p_in=13.19) has headroom vs the ensemble floor (p_in=12.1). Lion's deterministic sign quantization may trap the model in a sharp local minimum. Langevin noise forces exploration of nearby parameter space, potentially finding flatter basins that generalize better. The noise is completely orthogonal to the loss function — L1 gradient structure is fully preserved.

**Key constraint:** EMA only captures the last ~20 epochs. Noise must be near-zero well before epoch 140. We use a linear decay schedule: `noise_scale = cfg.langevin_noise * max(0, 1 - epoch / (cfg.cosine_T_max * 0.8))`.

**References:**
- Welling & Teh (2011) — SGLD: Bayesian Learning via Stochastic Gradient Langevin Dynamics
- Chen et al. (2016) — Bridging the Gap between Stochastic Gradient MCMC and Stochastic Optimization

## Instructions

**Step 1: Add argparse flag and Config field:**
```python
parser.add_argument('--langevin_noise', type=float, default=0.0,
                    help='Langevin noise scale added to params after Lion step. 0 = disabled.')
```
And in the Config dataclass:
```python
langevin_noise: float = 0.0  # SGLD-style parameter noise after Lion step
```

**Step 2: Add noise injection after the optimizer step.** Find the line that calls `optimizer.step()` (~line 1700+ in train.py). Immediately after `optimizer.step()` and BEFORE the EMA update, add:
```python
# Langevin noise for SGLD-style exploration (anneals to 0 before EMA window)
if cfg.langevin_noise > 0.0:
    noise_scale = cfg.langevin_noise * max(0.0, 1.0 - epoch / (cfg.cosine_T_max * 0.8))
    if noise_scale > 0.0:
        with torch.no_grad():
            for p in model.parameters():
                if p.requires_grad:
                    p.data.add_(torch.randn_like(p.data) * noise_scale)
```

That is all — ~10 LoC. No other changes needed.

**Step 3: Run 6 experiments** (3 noise scales × 2 seeds). All runs must include `--aft_foil_srf` (baseline since PR #2104).

W&B group: `phase6/langevin-noise`

noise=5e-5 (very conservative):
```bash
cd cfd_tandemfoil && python train.py --agent edward \
  --wandb_name "edward/langevin-5e5-s42" --wandb_group phase6/langevin-noise \
  --langevin_noise 5e-5 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf
```
Run seed 73 with the same flags (`--wandb_name "edward/langevin-5e5-s73" --seed 73`).

noise=1e-4 (primary test value):
Same flags with `--langevin_noise 1e-4`, names `edward/langevin-1e4-s42` and `edward/langevin-1e4-s73`.

noise=3e-4 (aggressive — monitor for instability):
Same flags with `--langevin_noise 3e-4`, names `edward/langevin-3e4-s42` and `edward/langevin-3e4-s73`.
If val/loss spikes above 0.41 during training, abort and skip the second seed for that value.

**Report all 6 runs** (or 5 if 3e-4 diverges): p_in, p_oodc, p_tan, p_re per run + W&B run IDs.

## Baseline

Current single-model baseline (PR #2104, 8-seed mean, seeds 42-49, +aft_foil_srf):

| Metric | 8-seed mean | Target to beat |
|--------|-------------|----------------|
| p_in   | **13.19 ± 0.33** | < 13.19 |
| p_oodc | **7.92 ± 0.17**  | < 7.92  |
| p_tan  | **30.05 ± 0.36** | < 30.05 |
| p_re   | **6.45 ± 0.07**  | < 6.45  |

Single-model seed references (no Langevin, with --aft_foil_srf):
- seed=42: p_in≈12.71, p_oodc≈8.13
- seed=73: p_in≈12.2, p_oodc≈7.59

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent edward \
  --wandb_name "edward/baseline-s42" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

**Update: Relaunched experiments with adjusted EMA timing**

The initial 6 runs were killed at epoch ~63 because I realized the default `ema_start_epoch=140` would not be reached within the 180-minute wall-clock timeout (at ~82s/epoch, timeout hits at ~epoch 132).

Relaunched all 6 experiments with `--ema_start_epoch 100`, giving ~32 epochs of EMA averaging before the timeout. Also launched 2 baseline runs (no noise, same `--ema_start_epoch 100`) on seeds 42 and 73 for apples-to-apples comparison.

Experiments now running on all 8 GPUs:
- 6 Langevin (3 noise scales × 2 seeds) + 2 baselines

W&B group: `phase6/langevin-noise`

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### All runs — key metrics (Surface MAE pressure)

| Run | p_in | p_oodc | p_tan | p_re | W&B run ID |
|-----|------|--------|-------|------|------------|
| **Baseline (no noise) s42** | **12.9** | 7.9 | **29.9** | 6.6 | vhz8bzct |
| **Baseline (no noise) s73** | 13.1 | 8.0 | 30.1 | 6.5 | ei43kzqz |
| noise=5e-5 s42 | 13.4 | **7.7** | 29.9 | 6.6 | cbm4a9fs |
| noise=5e-5 s73 | 13.3 | **7.7** | 30.0 | **6.4** | 13f5pbul |
| noise=1e-4 s42 | 13.0 | 7.9 | 30.9 | 6.5 | twgrbt24 |
| noise=1e-4 s73 | 13.8 | 8.0 | 30.2 | **6.3** | 3ao1faty |
| noise=3e-4 s42 | 13.4 | 7.9 | 31.2 | 6.4 | ff02he97 |
| noise=3e-4 s73 | 13.0 | 8.2 | **29.7** | 6.6 | namev902 |

### Averages by noise level

| Noise Scale | p_in (avg) | p_oodc (avg) | p_tan (avg) | p_re (avg) |
|-------------|------------|--------------|-------------|------------|
| Baseline (0) | **13.0** | 7.95 | **30.0** | 6.55 |
| 5e-5 | 13.35 | **7.70** | 29.95 | **6.50** |
| 1e-4 | 13.40 | 7.95 | 30.55 | **6.40** |
| 3e-4 | 13.20 | 8.05 | 30.45 | 6.50 |

### Baseline comparison (PR #2104, 8-seed mean with --aft_foil_srf)

| Metric | PR #2104 baseline | Our baseline (2 seeds) | Best Langevin |
|--------|-------------------|----------------------|---------------|
| p_in | 13.19 | 13.0 | 13.0 (5e-5 s42) |
| p_oodc | 7.92 | 7.95 | 7.7 (5e-5, both seeds) |
| p_tan | 30.05 | 30.0 | 29.7 (3e-4 s73) |
| p_re | 6.45 | 6.55 | 6.3 (1e-4 s73) |

### Experimental details

- All runs used `--ema_start_epoch 100` (adjusted from default 140 to ensure EMA activates before 180-min wall-clock timeout)
- Runs trained to epoch 147-150 before timeout (50 epochs of EMA averaging)
- Noise schedule: `noise * max(0, 1 - epoch / 128)` — noise fully annealed to 0 at epoch 128
- Baseline runs used identical config with `langevin_noise=0`
- Peak VRAM: ~89 GB per GPU

### Commands used

```bash
# Example (noise=5e-5, seed=42):
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py --agent edward \
  --wandb_name "edward/langevin-5e5-s42-v2" --wandb_group phase6/langevin-noise \
  --langevin_noise 5e-5 --seed 42 --ema_start_epoch 100 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf
```

### What happened

**Langevin noise had no meaningful positive effect on p_in or p_tan; marginal improvements on p_oodc and p_re within noise.**

The results are essentially within noise of the baseline. Key observations:

1. **p_in**: Baseline (13.0) is better than or equal to all Langevin runs. The noise slightly degrades p_in, particularly at 1e-4 (13.4).

2. **p_oodc**: The 5e-5 noise level shows a small improvement (7.7 vs 7.95), but this is within the typical seed variance (±0.2).

3. **p_tan**: All Langevin runs are comparable to baseline (~30.0). The 3e-4 s73 run hit 29.7 but this appears to be seed luck — the same noise level with s42 gave 31.2 (worst of all runs).

4. **p_re**: Small improvements at noise=1e-4 (6.3-6.5 vs 6.55), but again within noise.

5. **No instability at 3e-4**: The aggressive noise level did not cause divergence, but also did not help.

**Why it didn't work:**
- Lion's sign-based update already provides implicit gradient noise (all updates are ±lr regardless of gradient magnitude). Adding isotropic Gaussian noise on top is redundant.
- The noise-to-signal ratio at these scales (0.25x to 1.5x Lion's step size) is conservative, but even at 1.5x there's no exploration benefit — the EMA already smooths over per-step stochasticity.
- With 50 epochs of EMA averaging (ema_decay=0.999 over ~16,500 batches), any noise-induced exploration is completely averaged out in the final model.

### Suggested follow-ups

1. **Bernoulli noise instead of Gaussian**: The Sign-SGLD literature suggests Bernoulli ±1 noise is more natural for sign-based optimizers. This might interact more constructively with Lion's momentum.
2. **Reduced noise for refinement heads**: The researcher-agent suggested using 0.5x noise scale for the smaller refinement heads. Their proportionally larger perturbation might be hurting p_tan.
3. **Higher temperature with reduced EMA**: If the goal is flat-minima exploration, the noise needs to be much larger (~10x-100x) with a correspondingly earlier annealing cutoff, possibly combined with reduced EMA decay to retain more of the exploration diversity.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2120 because ADVISOR: Good experiment, edward. Your analysis is spot-on — Lion's sign operation already provides implicit gradient noise, making SGLD-style additive noise redundant. The EMA further smooths out any exploration benefit.

**Key findings logged:**
- Langevin noise at all 3 scales (5e-5, 1e-4, 3e-4) shows no meaningful improvement over the no-noise control baseline within this experiment
- p_oodc marginal improvement at 5e-5 (7.7 vs 7.95) is within typical seed variance (±0.2)
- p_tan shows high variance (29.7 to 31.2 at 3e-4) — noise adds instability without benefit
- Lion + EMA makes this entire direction redundant

**Closing as dead end.** The underlying insight (implicit gradient noise from Lion's sign operation) is valuable — it rules out the entire family of gradient perturbation approaches for this optimizer.

Also: good catch on the ema_start_epoch timing issue. That's useful information.

I'll assign you a new experiment shortly..
